src/layers/generic.cpp

```diff
@@ -462,6 +473,21 @@ namespace marian {
     E_ = graph_->param(name, {dimVoc, dimEmb}, initFunc, fixed);
   }
 
+  //Embeds a sequence of words (given as indices), where they have factor information. The matrices are concatenated
+  /*private*/ Expr Embedding::embedWithConcat(const Words& data) const {
+    auto graph = E_->graph();
+    std::vector<IndexType> lemmaIndices;
+    std::vector<float> factorIndices;
+    factoredVocab_->lemmaAndFactorsIndexes(data, lemmaIndices, factorIndices);
+    auto lemmaEmbs = rows(E_, lemmaIndices);
+    int dimFactors =  FactorEmbMatrix_->shape()[0];
+    auto factEmbs = dot(graph->constant({(int) data.size(), dimFactors}, inits::fromVector(factorIndices), Type::float32), FactorEmbMatrix_);
+
+    auto out = concatenate({lemmaEmbs, factEmbs}, -1);
+
+    return out;
+  }
+
```

Comment:
To embed using concatenation basically we call the lemmaAndFactorsIndexes() function that decodes each wordIndex into its lemma and factor information. Then we embed lemmas and factors separately. To embed the lemmas we just choose the rows of the lemma embedding matrix (E_) that correspond to the indexes of the lemmas in the vocab given by lemmaAndFactorsIndexes. Then to embed the factors, since that a token could have more than one factor, we cannot simple chose the rows of the factor embedding matrix as we did for the lemmas. So for the factors an array of 0's and 1's is returned, where for each token we have the information if a certain factor appears or not in a token. This vector is turned into a sparse matrix that is multiplied by the factor embedding matrix to get the factor embeddings. This two result embeddings (lemmaEmbs and factEmbs) are then concatenated.